### DISCUSSION TITLE
How complex is to add an "unsupported" OCPP charger?

### DISCUSSION THREAD

**Discussion by stefa168** [lbbrhzn/ocpp#542] [Q&A]
Hi, pretty soon I will get my first EV; I am choosing the charger and since my house has solar panels and a HA server the choice of the charger will be heavily influenced towards those that allow to be controlled in some way to limit charging exclusively with solar whenever possible.

Right now I am considering some models [from SCAME](e-mobility.scame.com) which, from what I've seen in this project's documentation, haven't been tested yet.

So this is my question: how hard is to add compatibility for an unsupported charger? Thanks!

**Comment by GuzziDoug** [comment]
Hi,
 can this request be answered please? I have obtained a ZJBeny OCPP charger (single phase) for evaluation. ZJBeny built the original Tesla Wallbox & now produce their own charger (also sold under other names, but the case is the same as the Tesla one I think).
 I have the charger running off my own SteVe server, but want to integrate it into my fairly complicated home PV setup. I am on 240v 2 phase rural in Australia, a 180 deg phase 240v-0-240v setup. I have 2 of Solaredge SE5000 inverters, & 2 of Victron Multiplus 2 inverters that are connected to abt 12Kw of Zenaji Lithium Titanate batteries. I have Node Red loaded onto my Victron Venus controller. I intend reading the data from the inverters using Node Red (due to distance between Generation area & loads), & the usage data from my IOTaWatt monitoring the house usage.
 What I hope to achieve is to integrate the above into Home Assistant so the charging & manageable loads can be managed to minimise the draw from the grid.
 In order to achieve this, I will need to integrate the ZJBeny charger into this project. Can anyone point me to the information on setting up a new charger please?

**Comment by lbbrhzn** [comment]
Hi Stefa168, GuzziDoug,

When a charger is not listed as a supported charger it simply means that it has not been reported to work. Whether it will work or not in practice really depends on whether it is compliant with the OCPP standard. Some vendors claim their device is compliant without bothering to do a compliance test, because that takes time and costs money! 

When it is fully compliant, then it should work out of the box, since the ocpp integration is designed to work for fully compliant chargers. Any issues should be reported, and we will do out best to analyze them. In some cases modifications or workarounds may be needed.  As long as these workarounds do not break compliance to the OCPP standard they can be added to this repository. 
Otherwise, we urge you to request your vendor to update their firmware to make their device OCPP compliant.

You can always make your own fork of this repository to solve issues for a specific device that are not OCPP compliant. However, we will not integrate these type of changes into this repository, because that may prevent other chargers to work.

Kind regards,

lbbrhzn

**Comment by BjarneBitscrambler** [comment]
I see @lbbrhzn replied while I was typing... :-)  Hopefully the following is of some use :-)

I haven't been involved in authoring or maintaining this project, so I'm not an expert...I base my following comments on my experience with connecting my Grizzl-E charger.

The difficulty level really depends on several factors:
1.  how compliant a particular charger is to the OCPP specification. 
2.  how many other folks are using the same model of charger
3.  how experienced you are at python programming
...and I should probably add a 4th factor:  your persistence :-)

With respect to point (1), a lot of manufacturers claim OCPP compliance, yet fall short in practice. I've read the spec, and it's pretty detailed and clear, but I suspect most time-limited designers only bother to implement and test a sub-set of the requirements. As an example, United Chargers' Grizzl-E works fine with the ChargeLabs app that they recommend, but reboots continuously when trying to communicate with HA OCPP server. It's not the fault of the HA-OCPP; the Grizzl-E responds with invalid JSON to several OCPP messages and kacks out. ChargeLabs OCPP server doesn't use these messages, so this functionality was not tested by the manufacturer.

So, it's hard to tell whether a charger will work out-of-the-box with HA-OCPP without actually hooking it up and testing it. I'd suggest that's your first step.

If it doesn't work, then point (2) comes into play. Perhaps someone else has figured out what's going wrong, and may even have come up with a workaround.  For the Grizzl-E there was a long thread of contributions (#442) by various folks to eventually get it mostly working.

However, if you are at the bleeding edge and are the first to try your model of charger you have some detective work to do. By enabling HA-OCPP debug messages and using a packet-sniffer (e.g. *WireShark* or *Burp Suite*) to examine the websocket messages between the server and the charger, you can figure out where the deviations from the OCPP spec occur.  If it's the fault of HA-OCPP, then there's a big likelihood that @lbbrhzn or another maintainer will fix it.  If it's the fault of the charger, then either you will have to get the manufacturer to fix their firmware, or you will have to tweak the python code of the OCPP server to workaround the issue and live with the limitations. Back to my charger example - we bypass the several lines in the server's code that sends out the messages that the Grizzl-E can't handle and it works. 

Good luck!

**Comment by GuzziDoug** [comment]
Thanks! I am currently loading Mint, to load Home assistant (etc). This charger runs fine with SteVe, so (cross fingers) should be OK.

 I will report back....

 FYI, this charger is very well priced (in Australia). The OCPP model seems to be a new release of this charger. My interest came because I was a voluntary Energy Coach (as a retired Tech) with Enova Energy until this community owned energy reseller was sent broke by the big Gentaillers (who own both generation & retailer businesses). The OCPP charger was to be an Energy shedding project for Enova (in exchange for a better EV charge rate.)  I still think this EV load shedding is required for a stable renewable grid (so OCPP chargers should perhaps be mandated)

Thanks again.
